src/cryptography.js

```diff
@@ -0,0 +1,189 @@
+ * session.
+ *
+ * The wrapping and data algorithm suites are chosen for the following
+ * properties:
+ *
+ *   - Symmetric encryption using AES-GCM with 256-bit keys (AES256_GCM_IV12_TAG16)
+ *   - Data key derivation with HKDF and SHA-384 (HKDF_SHA384)
+ *   - No signing (ECDSA_*)
+ *   - No key commitments (COMMIT_KEY)
+ *
+ * We opt out of the defaults of signing and key commitments because they both
+ * add significant overhead to the encrypt/decrypt time and (to a lesser
+ * degree) the encrypted message size.  Since we decrypt on ~every request with
+ * a session, it's important to be as fast as possible!  These security
+ * features are also not exceptionally useful for our use case, as they
+ * primarily protect against modifications of the ciphertext by external
+ * parties but we're not exchanging the data with anyone other than ourselves.
+ * In rough benchmarking, the encrypt() and decrypt() routines provided by this
+ * module with the algorithm choices above each take <1ms.
+ *
+ * See the AWS Cryptography SDK docs¹ for more information on these properties.
+ * The concept guide² is a particularly helpful place to start and get oriented
+ * before reading further reference material.
+ *
+ * ¹ https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/
+ * ² https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html
+ *
+ * @module cryptography
+ */
+
+const assert = require("assert").strict;
+const awsCrypto = require("@aws-crypto/client-node");
+const {randomBytes} = require("crypto");
+
+
+/* These must be changed with care.  Blithely changing them could make existing
+ * ciphertexts undecryptable.
+ */
+const KEY_LENGTH_BITS = 256;
+const KEY_COMMITMENT_POLICY = awsCrypto.CommitmentPolicy.FORBID_ENCRYPT_ALLOW_DECRYPT;
+const WRAPPING_SUITE = awsCrypto.RawAesWrappingSuiteIdentifier.AES256_GCM_IV12_TAG16_NO_PADDING;
+const DATA_SUITE = awsCrypto.AlgorithmSuiteIdentifier.ALG_AES256_GCM_IV12_TAG16_HKDF_SHA256;
+
+
+const {
+  encrypt: _encrypt,
+  decrypt: _decrypt,
+} = awsCrypto.buildClient(KEY_COMMITMENT_POLICY);
+
+
+/**
+ * Construct a keyring from one or more AES 256-bit keys.
+ *
+ * The first key specified will be used for encryption.  All keys will be
+ * considered (in order) for decryption.  This allows for key rotation over
+ * time.
+ *
+ * @param {String} keyParamString - A URL query param string encoding pairs of
+ *    key names and base64-encoded key material
+ * @returns {Object}
+ */
+function keyringFromParamString(keyParamString) {
+  const keyParams = Array.from((new URLSearchParams(keyParamString)).entries());
+
+  // Decryption with any key passed in.
+  const children = keyParams.map(([keyName, encodedKey]) => {
+    assert(typeof encodedKey === "string");
+
+    const keyBytes = new Uint8Array(Buffer.from(encodedKey, "base64"));
+
+    assert(keyBytes.byteLength * 8 === KEY_LENGTH_BITS, "invalid key size");
```

Comment:
Is the purpose of these asserts to give immediate feedback if someone sets a bad key within `SESSION_ENCRYPTION_KEYS`? Just wondering why not just filter out the invalid keys.

Reply:
Yep, to give an immediate error. Filtering them out (silently or otherwise) would easily mask things like botched key rotations, e.g. I update `SESSION_ENCRYPTION_KEYS` to add a new key but do it wrong and the new key is filtered out for being invalid, so the key rotation doesn't happen but I think it did. In security-conscious code, it's good practice to be very defensive and abort/error rather than continue if things aren't exactly as they should be.